GafferUI::RampPlugValueWidget : Added SplineBasisWidget

danieldresser-ie opened the pull request:
This allows choosing the spline basis.  This should work along side the Cortex changes I'm working on to offer a better experience with the ramp editor ( note that in particular, the linear basis is quite broken without the Cortex fixes )

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -49,6 +49,9 @@ def __init__( self, plug, **kw ) :
 
 		with column :
+			with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 ) :
+				self.__basisLabel = GafferUI.LabelPlugValueWidget( plug["basis"] )
```
Let's ditch the member variables since they're never used.

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -49,6 +49,9 @@ def __init__( self, plug, **kw ) :
+			with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 ) :
+				self.__basisLabel = GafferUI.LabelPlugValueWidget( plug["basis"] )
+				self.__basisField = GafferUI.SplineBasisWidget( plug["basis"], plug["endPointMultiplicity"])
 
```
Could you try adding a spacer here, so that the menu doesn't dominate the layout quite so much? I think `GafferUI.Spacer( IECore.V2i( 0 ), parenting = { "expand" : True } )` might do it...

johnhaddon commented on python/GafferUI/SplineBasisWidget.py:
```diff
@@ -0,0 +1,176 @@
+		self.__readOnly = False
+
+		self.__presets = {
+			"linear" : ( IECore.CubicBasisf.linear(), 1 ),
```
Let's capitalise these since they appear in the UI.

johnhaddon commented on python/GafferUI/SplineBasisWidget.py:
```diff
@@ -0,0 +1,176 @@
+				if basisMatrix == values[0].matrix:
+					matrixName = label
+	
+				if basisMatrix == values[0].matrix and basisStep == values[0].step  and endPointMultiplicity == values[1]:
```
Minor thing, but we should just be able to compare CubicBasis objects directly rather than having to compare the step and matrix separately shouldn't we?

danieldresser-ie replied:
Currently, there's no direct way to get a CubicBasis from the basis plug, because it's a bare ValuePlug, which doesn't implement getValue().  I actually have a commit lying around where I add a SplineBasis plug type, so we could call getValue() on it, but I dropped this because it was extra code and didn't seem to add too much value.  I can add that back if you think it's valuable.

Or were you just suggesting I should temporarily construct a new CubicBasis from the two plug values inside the compare?

johnhaddon commented on python/GafferUI/SplineBasisWidget.py:
```diff
@@ -0,0 +1,176 @@
+		) :
+			self._updateFromPlugs()
+
+	def _currentLabel( self ) :
```
Let's make this private.

johnhaddon commented on python/GafferUI/SplineBasisWidget.py:
```diff
@@ -0,0 +1,176 @@
+
+		return text
+
+	def _updateFromPlugs( self ) :
```
Let's make this private too.

johnhaddon commented on python/GafferUI/SplineBasisWidget.py:
```diff
@@ -0,0 +1,176 @@
+			basisName = None
+			matrixName = "Unknown"
+
+			with self.__basisPlug.ancestor( Gaffer.ScriptNode.staticTypeId() ).context():
```
If this were a PlugValueWidget, we'd have `self.getContext()` here. It is feeling like we're duplicating a lot of the PlugValueWidget functionality and maybe we'd be better off just deriving from it - I don't recall the details of our last discussion, but maybe it's my fault and I suggested otherwise?

danieldresser-ie replied:
Yes, you did suggest otherwise.  It felt kind of dirty to me to be a PlugValueWidget for basisPlug but implicitly depend equally much on endPointMultiplicityPlug.

Making it a PlugValueWidget for a SplinePlug seems like a potentially nice alternative though.  Maybe I should take a whack at that.

johnhaddon commented on python/GafferUI/SplineBasisWidget.py:
```diff
@@ -0,0 +1,176 @@
+		self.__menuButton = GafferUI.MenuButton( "", menu = GafferUI.Menu( Gaffer.WeakMethod( self.__menuDefinition ) ) )
+		GafferUI.Widget.__init__( self, self.__menuButton, **kw )
+
+		# We only set up signals based on the node of one plug - the two plugs should always be on the same node
```
It feels a bit wrong to provide a constructor that makes it possible to break this assumption. Maybe the constructor should just be passed the basis plug and it should find the endPointMultiplicityPlug? Or maybe the constructor should just take a SplinePlug? If we did the latter, I think deriving from PlugValueWidget might possibly work quite naturally, since `_updateFromPlug()` would be called whenever either child changed...

danieldresser-ie commented:
I think I've addressed all your comments, except for moving it to being private.  Now that it is a PlugValueWidget for the spline plug, it is a lot cleaner.  Are you still concerned by it being public?

I've also included a commit here which I could split into it's own PR if you want me to:  while working on the Cortex code to try and make the handles line up as well as possible with the resulting curve, I noticed that the method of drawing ramps was introducing some noticeable misalignment between the handles and the curve.  It's a bit hard to describe, and I'm not sure if the same thing would happen on OS X, but the gradient evaluation gets quantized as it approaches the right side of the widget, so that popping up a spline widget and scaling it to near the width of the screen would show the handles scaling smoothly, but any discontinuities in the ramp would jump in large increments.  Bizarrely, this quantization reset around 2048 pixels, so if I started dragging it onto the next monitor over, there would be a region where the ramp behaved smoothly before starting to become quantized again.

I can try and video capture this behaviour if you need to see it, but QLinearGradient seems like the wrong tool for the job anyway.  I would expect Qt to be better optimized for blitting scaled images than rendering linear gradients with 500 samples.  Since the switch, it seems to behave smoothly and line up perfectly.

johnhaddon commented:
Thanks Daniel - I've renamed the SplineBasisWidget to reflect its new base class (but left it public) and merged. I was able to reproduce some odd QLinearGradient behaviour on OSX too, which is also improved by the fix you've made.
Cheers...
John
